### PR TITLE
15 board score

### LINKED ISSUE #15: Board score
Determine who is ahead based on captured pieces.
Return positive number if white is winning.

image

### PR DESCRIPTION
**Pull request by chanson02**

### CODE REVIEW

File: src/board/evaluate.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        score += 5 * (self.white_rooks.count_ones() as i32 - self.black_rooks.count_ones() as i32);
+        score += 9 * (self.white_queen.count_ones() as i32 - self.black_queen.count_ones() as i32);
+
+        // TODO: Take into account amount of legal moves per piece
+        score
+    }
+
+    // I would like to clean up the nesting in here --Cooper
+    /// This function does not validate that there is a pawn at this position
+    pub fn get_legal_pawn_moves(&self, from: i64, white: bool) -> Vec<i64> {
```
Comment (chanson02) [inline src/board/evaluate.rs]: change this to `get_pawn_moves`

File: src/board/movement.rs [hunk 2]
```diff
@@ -60,6 +60,15 @@ impl Chessboard {
         None
     }
 
+    /// Checks to see if there is a white piece at this position
+    ///
+    /// # Returns
+    ///
+    /// true if white, black is false, none if nothing
+    pub fn white_at_position(&self, square: i64) -> Option<bool> {
+        Some(self.piece_at_position(square)?.is_ascii_uppercase())
```
Comment (chanson02) [inline src/board/movement.rs]: change this to use the new `black_side_pieces` which is faster.
https://github.com/chanson02/KnO3/pull/79

File: engine/src/game_state.rs [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+            "-" => 255, // out of range
+            _ => position::string_to_square(parts[3])?,
+        };
+
+        Ok(Self {
+            white_turn: parts[1] == "w",
+            castling: parse_castling_rights(parts[2]),
+            en_passant: passant,
+            half_clock: parts[4].parse().unwrap(),
+            move_count: parts[5].parse().unwrap(),
```
Comment (chanson02) [inline engine/src/game_state.rs]: I wonder if we could do something safer here.

File: README.md [hunk 4]
```diff
@@ -1,22 +1,22 @@
 # KnO3 - A Rusty Knight Chess Engine
 
-This project is a chess engine built in rust. It provides an engine for playing chess, and can be attached to any GUI.
-
+This project is a chess engine built in rust. It provides an engine for playing chess, and can be attached to any GUI
 ## Primary Contributors
 - [Cooper Hanson](https://github.com/chanson02/)
```
Comment (abrahimt) [inline README.md]: nit: remove the slash after your username. It's not needed. Major code smell

File: engine/src/position.rs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        if bitboard & 1 != 0 {
+            squares.push(index);
+        }
+        index += 1;
+        bitboard >>= 1;
+    }
+    squares
+}
+
+#[cfg(test)]
```
Comment (abrahimt) [inline engine/src/position.rs:58]: Why are the tests not actually being run? cargo test runs 0 tests?

File: cli/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Error::FENParsingError(msg) => write!(f, "FEN Parsing error: {}", msg),
+            Error::ArgumentError(msg) => write!(f, "Argument error: {}", msg),
+        }
+    }
+}
+
+fn main() -> Result<(), Error> {
+    let matches = Command::new("KnO3 Chess CLI")
```
Comment (abrahimt) [inline cli/src/main.rs:24]: Can we put this somewhere else? Not super familiar with adding commands

File: engine/src/game_state.rs [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+            'k' => 0b0010,
+            'q' => 0b0001,
+            _ => 0,
+        };
+        result |= v;
+    }
+    result
+}
+
+#[cfg(test)]
```
Comment (abrahimt) [inline engine/src/game_state.rs:108]: I don't think these are actually getting run

File: engine/src/move_generation.rs [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+            'r' => self.possible_rook_moves(square, is_white),
+            'n' => self.possible_knight_moves(square, is_white),
+            'b' => self.possible_bishop_moves(square, is_white),
+            'k' => self.possible_king_moves(square, is_white),
+            'q' => self.possible_queen_moves(square, is_white),
+            _ => vec![],
+        }
+    }
+
+    // TODO: en passant check
```
Comment (abrahimt) [inline engine/src/move_generation.rs]: This still a todo? We should probably get that done before merging. I can do it if you want

File: engine/src/move_generation.rs [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+        result
+    }
+
+    fn possible_queen_moves(&self, from: u8, white: bool) -> Vec<u8> {
+        let mut result = self.possible_rook_moves(from, white);
+        result.extend(self.possible_bishop_moves(from, white));
+        result
+    }
+
+    // TODO: Make sure they are not moving into check/mate
```
Comment (abrahimt) [inline engine/src/move_generation.rs]: Let's get these todos done

File: cli/src/main.rs [hunk 10]
```diff
@@ -11,7 +11,6 @@ enum Error {
     ArgumentError(String),
 }
 
-// I don't think this is doing anything but it's suppressing a warning -Cooper
```
Comment (abrahimt) [inline cli/src/main.rs]: We obviously all know what's happening here. Good call removing the comment!

File: engine/src/move_generation.rs [hunk 11]
```diff
@@ -246,44 +245,37 @@ impl GameState {
...
-        let east = horz & greater_than; //king side
-        let west = horz & less_than; //queen side
-        
-        // todo: think about using horz to make black white neutral if statements
-        if east & 0b0110000 == 0 && (capital_k && white || k && !white) {
-            filtered |= 0b01000000;
+        // check which side we can castle too
+        let king_side = capital_k & white || k & !white;
+        let queen_side = capital_q & white || q & !white;
+        if (!king_side & !queen_side) || self.position_under_attack(from, white) { return filtered; }
```
Comment (abrahimt) [inline engine/src/move_generation.rs]: Beautiful.

File: engine/src/move_generation.rs [hunk 12]
```diff
@@ -246,44 +245,37 @@ impl GameState {
...
+        let west = horz & less_than;
+
+        let blocked_k_side = !king_side || east & (0b01100000 << shift) != 0;
+        let blocked_q_side = !queen_side || west & (0b00001110 << shift) != 0;
+        if !blocked_k_side && !self.position_under_attack(from + 1, white) && !self.position_under_attack(from + 2, white) {
+            filtered |= 0b01000000 << shift;
         }
-        if west & 0b00001110 == 0 && (capital_q && white || q && !white) {
-            filtered |= 0b00000100;
+        if !blocked_q_side && !self.position_under_attack(from - 1, white) && !self.position_under_attack(from - 2, white) {
```
Comment (abrahimt) [inline engine/src/move_generation.rs]: checks out

**Review by abrahimt — DISMISSED** [review]
I did a lot of this so I trust it.

File: cli/src/main.rs (hunk 6 above)
Comment (chanson02) [reply cli/src/main.rs:24]: We should make an issue for constructing how we want this CLI to work.

File: README.md (hunk 4 above)
Comment (abrahimt) [reply README.md]: fixed it myself

File: cli/src/main.rs (hunk 6 above)
Comment (abrahimt) [reply cli/src/main.rs:24]: Agreed.

**Review by abrahimt — APPROVED** [review]

**Pull request merged**